### ISSUE TITLE
Discussion: Update dataloader to skip rows that dont require training

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#2344]
https://github.com/pytorch/torchtune/issues/2341

When a) `train_on_input=False` and b) message is too long that output is truncated, there may be a batch without trainable tokens, raising an error on the loss because of division by zero.

Beyond raising an inconvenient bug, this is a waste of compute.

In the dataloader, should we skip rows that does have trainable embeddings?

**Comment by RdoubleA** [comment]
It's difficult to know if this is the case fro the DataLoader perspective. You would have to catch this and raise/skip in the recipe or add additional preprocessing to your dataset to prevent this. We don't have the ability to skip an item within the `SFTDataset`'s `__getitem__`.

**Comment by felipemello1** [comment]
I see. It could be added to the recipe, like "check_batch_requires_grad(batch)". I am sure @ebsmothers will love the idea!

**Comment by EugenHotaj** [comment]
Maybe slightly orthogonal but it would also be great to expose a way to do left truncation instead of right truncation. I think this is even more important once the "mask all previous turns" strategy is introduced.

(I'm actually a bit surprised to see that torchtune supports left truncation only since I thought right truncation was the standard practice for llama models).

**Comment by pocca2048** [comment]
> You would have to catch this and raise/skip in the recipe

I agree. Actually, I implemented skipping logic in the recipe w/o fixing the loss.
